### PR TITLE
[Feature] Implement automatic partition creation during data loading

### LINKED ISSUE #17935: Automatic partition creation during data ingestion
### Create Automatic Partition Table
```
CREATE TABLE ss(
    event_day DATE,
    pv BIGINT
)
DUPLICATE KEY(event_day)
PARTITION BY date_trunc('month', event_day) 
DISTRIBUTED BY HASH(event_day) BUCKETS 8
PROPERTIES("replication_num" = "3");
```
### Insert Values
```
insert into ss values('2023-02-14', 2),('2023-03-01',2);
```

### PR DESCRIPTION
**Pull request by meegoo**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #17935

#### Create Automatic Partition Table
```
CREATE TABLE ss(
    event_day DATE,
    pv BIGINT
)
DUPLICATE KEY(event_day)
PARTITION BY date_trunc('month', event_day) 
DISTRIBUTED BY HASH(event_day) BUCKETS 8
PROPERTIES("replication_num" = "3");
```

#### Insert Values
```
insert into ss values('2023-02-14', 2),('2023-03-01',2);
```

#### Show Partitions
```
mysql> show partitions from ss;
+-------------+---------------+----------------+---------------------+--------------------+--------+--------------+-------+-----------------+---------+----------------+---------------+---------------------+--------------------------+----------+------------+----------+
| PartitionId | PartitionName | VisibleVersion | VisibleVersionTime  | VisibleVersionHash | State  | PartitionKey | Range | DistributionKey | Buckets | ReplicationNum | StorageMedium | CooldownTime        | LastConsistencyCheckTime | DataSize | IsInMemory | RowCount |
+-------------+---------------+----------------+---------------------+--------------------+--------+--------------+-------+-----------------+---------+----------------+---------------+---------------------+--------------------------+----------+------------+----------+
| 14042       | p202302       | 2              | 2023-02-15 11:01:12 | 0                  | NORMAL |              |       | event_day       | 8       | 3              | HDD           | 9999-12-31 23:59:59 | NULL                     | 0B       | false      | 0        |
| 14075       | p202303       | 2              | 2023-02-15 11:01:12 | 0                  | NORMAL |              |       | event_day       | 8       | 3              | HDD           | 9999-12-31 23:59:59 | NULL                     | 0B       | false      | 0        |
+-------------+---------------+----------------+---------------------+--------------------+--------+--------------+-------+-----------------+---------+----------------+---------------+---------------------+--------------------------+----------+------------+----------+
2 rows in set (0.01 sec)
```

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by meegoo** [comment]
run starrocks_admit_test

File: gensrc/proto/internal_service.proto [hunk 1]
```diff
@@ -164,6 +164,7 @@ message PTabletWriterOpenRequest {
     optional int64 timeout_ms = 24;
     optional WriteQuorumTypePB write_quorum = 25;
     optional string merge_condition = 26;
+    optional bool is_incremental = 27;
```
[the remark this answers is not in the document]
Comment (imay) [reply gensrc/proto/internal_service.proto:167]: open lazy or late?

**Review by imay — DISMISSED** [review]
Must keep the same ordinal in Thrift/Protobuf definition files.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 2 / 20 (10.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|18|00.00%|[1563, 1569, 1574, 1581, 1587, 1596, 1602, 1608, 1619, 1625, 1634, 1640, 1646, 1655, 1660, 1671, 1682, 1683]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|1|100.00%|[]|

**Comment by meegoo** [comment]
>

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 2 / 20 (10.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|0|18|00.00%|[1563, 1569, 1574, 1581, 1587, 1596, 1602, 1608, 1619, 1625, 1634, 1640, 1646, 1655, 1660, 1671, 1682, 1683]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 3 / 21 (14.29%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|1|19|05.26%|[1565, 1571, 1576, 1583, 1589, 1598, 1604, 1610, 1621, 1627, 1636, 1642, 1648, 1657, 1662, 1673, 1684, 1685]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|1|100.00%|[]|

File: be/src/exec/tablet_sink.cpp [hunk 2]
```diff
@@ -1149,9 +1262,39 @@ Status OlapTableSink::send_chunk(RuntimeState* state, Chunk* chunk) {
...
+                _partition_not_exist_row_values.clear();
+                // only support single column partition now
+                _partition_not_exist_row_values.resize(1);
+                RETURN_IF_ERROR(_vectorized_partition->find_tablets(chunk, &_partitions, &_tablet_indexes,
+                                                                    &_validate_selection, &invalid_row_index, _txn_id,
+                                                                    &_partition_not_exist_row_values));
+
+                if (!_partition_not_exist_row_values[0].empty()) {
+                    _is_automatic_partition_running.store(true, std::memory_order_release);
+                    RETURN_IF_ERROR(_automatic_partition_token->submit_func([this] {
```
Comment (chaoyli) [inline be/src/exec/tablet_sink.cpp:1276]: If the ingestion failed, the partition will be rollbacked or not?

File: be/src/exec/tablet_sink.cpp [hunk 3]
```diff
@@ -1095,21 +1123,106 @@ bool OlapTableSink::is_full() {
...
+    TCreatePartitionRequest request;
+    TCreatePartitionResult result;
+    request.__set_txn_id(_txn_id);
+    request.__set_db_id(_vectorized_partition->db_id());
+    request.__set_table_id(_vectorized_partition->table_id());
+    request.__set_partition_values(_partition_not_exist_row_values);
+    TNetworkAddress master_addr = get_master_address();
+    RETURN_IF_ERROR(ThriftRpcHelper::rpc<FrontendServiceClient>(
+            master_addr.hostname, master_addr.port,
+            [&request, &result](FrontendServiceConnection& client) { client->createPartition(result, request); }));
```
Comment (chaoyli) [inline be/src/exec/tablet_sink.cpp:1139]: If the master change during this period, it will failed.

File: be/src/runtime/exec_env.cpp [hunk 4]
```diff
@@ -162,6 +162,13 @@ Status ExecEnv::_init(const std::vector<StorePath>& store_paths) {
     _udf_call_pool = new PriorityThreadPool("udf", config::udf_thread_pool_size, config::udf_thread_pool_size);
     _fragment_mgr = new FragmentMgr(this);
 
+    RETURN_IF_ERROR(ThreadPoolBuilder("dynamic_partition") // dynamic partition pool
```
Comment (chaoyli) [inline be/src/runtime/exec_env.cpp:165]:
```suggestion
    RETURN_IF_ERROR(ThreadPoolBuilder("automatic_partition") // dynamic partition pool
```

File: gensrc/thrift/StatusCode.thrift [hunk 5]
```diff
@@ -100,6 +100,8 @@ enum TStatusCode {
     TXN_NOT_EXISTS = 51,
     TXN_IN_PROCESSING = 52,
 
-    RESOURCE_BUSY = 53
+    RESOURCE_BUSY = 53,
+
+    AUTOMATIC_PARTITION = 54
```
Comment (chaoyli) [inline gensrc/thrift/StatusCode.thrift:105]:
```suggestion
    LATE_PARTITION = 54
```
like `LATE MATERIALIZATION`

File: be/src/exec/tablet_sink.cpp (hunk 2 above)
Comment (meegoo) [reply be/src/exec/tablet_sink.cpp:1276]: not

File: be/src/exec/tablet_sink.cpp (hunk 3 above)
Comment (meegoo) [reply be/src/exec/tablet_sink.cpp:1139]: rpc may be failed by many reason, it doesn't matter.

File: gensrc/thrift/StatusCode.thrift (hunk 5 above)
Comment (meegoo) [reply gensrc/thrift/StatusCode.thrift:105]: It is different from the semantics of `LATE`, I changed it to `AUTOMATIC_PARTITION_IN_PROCESSING`

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 18 / 22 (81.82%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|16|20|80.00%|[1590, 1673, 1684, 1685]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 18 / 22 (81.82%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|16|20|80.00%|[1590, 1673, 1684, 1685]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|1|100.00%|[]|

**Review by Astralidea — DISMISSED** [review]

File: gensrc/thrift/StatusCode.thrift [hunk 6]
```diff
@@ -100,6 +100,8 @@ enum TStatusCode {
     TXN_NOT_EXISTS = 51,
     TXN_IN_PROCESSING = 52,
 
-    RESOURCE_BUSY = 53
+    RESOURCE_BUSY = 53,
+
+    AUTOMATIC_PARTITION_IN_PROCESSING = 54
```
[the remark this answers is not in the document]
Comment (meegoo) [reply gensrc/thrift/StatusCode.thrift:105]: Just use `EAGAIN` may be better if we want to reuse

File: be/src/runtime/local_tablets_channel.cpp [hunk 7]
```diff
@@ -578,6 +597,84 @@ StatusOr<std::shared_ptr<LocalTabletsChannel::WriteContext>> LocalTabletsChannel
...
+        tablet_ids.emplace_back(tablet.tablet_id());
+    }
+    _s_tablet_writer_count += _delta_writers.size();
+
+    std::stringstream ss;
+    ss << "open delta writer ";
+    for (auto& [tablet_id, delta_writer] : _delta_writers) {
+        ss << "[" << tablet_id << ":" << delta_writer->replica_state() << "]";
+    }
+    LOG(INFO) << ss.str();
```
[the remark this answers is not in the document]
Comment (meegoo) [reply be/src/runtime/local_tablets_channel.cpp:663]: Such logs are very useful for troubleshooting problems on the user side. Before that, the related logs were one line for each tablet, and now they have been reduced.

File: be/src/runtime/load_channel.cpp [hunk 8]
```diff
@@ -109,6 +110,15 @@ void LoadChannel::open(brpc::Controller* cntl, const PTabletWriterOpenRequest& r
             if (st = channel->open(request, _schema); st.ok()) {
                 _tablets_channels.insert({index_id, std::move(channel)});
             }
+        } else if (request.is_incremental()) {
+            LOG(WARNING) << "auto:incremental_open invoke";
+            // although shared_ptr's use_count is approximate in multithreaded environment
+            // but we protect shared_ptr ref by _lock
+            while (it->second.use_count() != 1) {
+                LOG(INFO) << "count " << it->second.use_count();
```
[the remark this answers is not in the document]
Comment (meegoo) [reply be/src/runtime/load_channel.cpp:118]: will not be a dead-loop but log too frequency

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 3 / 21 (14.29%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/service/FrontendServiceImpl.java|1|19|05.26%|[1598, 1604, 1609, 1616, 1622, 1631, 1637, 1643, 1654, 1660, 1669, 1675, 1681, 1690, 1695, 1706, 1717, 1718]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/planner/OlapTableSink.java|1|1|100.00%|[]|

**Review by Astralidea — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Pull request merged**